Source/WebCore/page/Frame.cpp

```diff
@@ -39,11 +39,64 @@
 namespace WebCore {
 
 #if ASSERT_ENABLED
-static HashMap<FrameIdentifier, WeakRef<Frame>>& allFrames()
-{
-    static NeverDestroyed<HashMap<FrameIdentifier, WeakRef<Frame>>> map;
-    return map;
-}
+class FrameLifetimeVerifier {
+public:
+    static FrameLifetimeVerifier& singleton()
+    {
+        static NeverDestroyed<FrameLifetimeVerifier> instance;
+        return instance.get();
+    }
+
+    void frameCreated(Frame& frame)
+    {
+        auto& pair = m_map.ensure(frame.frameID(), [] {
+            return std::pair<WeakPtr<LocalFrame>, WeakPtr<RemoteFrame>> { };
+        }).iterator->value;
+
+        switch (frame.frameType()) {
+        case Frame::FrameType::Local:
+            ASSERT_WITH_MESSAGE(!pair.first, "There should never be two LocalFrames with the same ID in the same process");
+            pair.first = downcast<LocalFrame>(frame);
```

Comment:
Since we already to type check for insertion, it seems we could just store two `Frame`s without downcasting:
`HashMap<FrameIdentifier, std::pair<WeakPtr<Frame>, WeakPtr<Frame>>> m_map;`

Reply:
We could do that, but this data structure makes it more clear what the exact type of each is and why we have a pair, and if we get anything wrong then down casting will fail.